### PR TITLE
bug: `RateCounter` rewrite

### PR DESCRIPTION
**Pull request by pmnoxx**
As @mm-near noticed, `truncate` method is not called before getting bandwidth stats. This can lead to displayed bandwidth data to be incorrect, especially if no data is send/received.

We will also refactor this code: https://github.com/near/nearcore/pull/5218

Edit:
- Rewrote the code completely from scratch, this will help us improve code quality and get rid of license.

See #5423

### CODE REVIEW

File: chain/network/src/peer/peer_actor.rs [hunk 1]
```diff
@@ -1107,6 +1107,13 @@ impl Handler<QueryPeerStats> for PeerActor {
     fn handle(&mut self, msg: QueryPeerStats, _: &mut Self::Context) -> Self::Result {
         #[cfg(feature = "delay_detector")]
         let _d = DelayDetector::new("query peer stats".into());
+
+        // Truncate to fix reporting in cases there is no data read/written.
+        // TODO(#5218) Refactor this code to use `SystemTime`
+        let now = millis_since_epoch();
+        self.tracker.sent_bytes.truncate(now);
```
Comment (mm-near) [inline chain/network/src/peer/peer_actor.rs:1114]: I don't like this solution - as it makes the API to rate counter quite confusing ("and remember to call the 'truncate' before calling bytes_per_minute if you didn't send any bytes recently")...

could we fix it within rate_counter.rs instead (by possibly checking the last time when truncate was run and running it from there?)

**Review by mm-near — CHANGES_REQUESTED** [review]

File: chain/network/src/peer/peer_actor.rs (hunk 1 above)
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs]: @mm-near I rewrite to change, to get and update at the same time.

File: chain/network/src/peer/peer_actor.rs [hunk 2]
```diff
@@ -1107,15 +1111,19 @@ impl Handler<QueryPeerStats> for PeerActor {
...
+        let (sent_per_min, sent_count_per_min) =
+            self.tracker.sent_bytes.get_bytes_per_min_and_count_per_min_and_truncate();
+        let (recv_per_min, recv_count_per_min) =
+            self.tracker.sent_bytes.get_bytes_per_min_and_count_per_min_and_truncate();
+
         PeerStatsResult {
             chain_info: self.chain_info.clone(),
-            received_bytes_per_sec: self.tracker.received_bytes.bytes_per_min() / 60,
-            sent_bytes_per_sec: self.tracker.sent_bytes.bytes_per_min() / 60,
+            received_bytes_per_sec: recv_count_per_min,
```
Comment (mm-near) [inline chain/network/src/peer/peer_actor.rs:1123]: shouldn't you also divide by 60 (as one is per-min, the other one per-sec)

**Review by mm-near — CHANGES_REQUESTED** [review]

File: chain/network/src/peer/peer_actor.rs [hunk 3]
```diff
@@ -1107,15 +1111,19 @@ impl Handler<QueryPeerStats> for PeerActor {
...
+            sent_bytes_per_sec: sent_count_per_min,
             is_abusive: self.is_abusive(),
-            message_counts: (
-                self.tracker.sent_bytes.count_per_min(),
-                self.tracker.received_bytes.count_per_min(),
-            ),
+            message_counts: (sent_per_min, recv_per_min),
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs]: Fixed.

File: chain/network/src/peer/peer_actor.rs (hunk 2 above)
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs]: Fixed

**Comment by pmnoxx** [comment]
@mm-near I added a few more code quality improvements.

**Comment by pmnoxx** [comment]
@matklad This is a total rewrite, you may want take a look.

**Comment by pmnoxx** [comment]
@matklad Can you take a look?

**Review by matklad — APPROVED** [review]

File: chain/network/src/peer/rate_counter.rs [hunk 4]
```diff
@@ -1,86 +1,91 @@
...
+    pub fn record(&mut self, bytes: u64, instant: Instant) {
+        self.bytes_per_min += bytes;
+        self.events.push_back(Event { instant, bytes })
     }
 
-    pub fn count_per_min(&self) -> u64 {
-        self.entries.len() as u64
+    // Get MinuteStats
+    // TODO(#5245) Remove instant once we add time mocking
+    pub fn minute_stats(&mut self, instant: Instant) -> MinuteStats {
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs:36]: Yes, let's keep it.

File: chain/network/src/peer/rate_counter.rs [hunk 5]
```diff
@@ -1,86 +1,91 @@
...
+}
 
-    pub fn bytes_per_min(&self) -> u64 {
-        self.bytes_sum
+impl TransferStats {
+    // Record event
+    // TODO(#5245) Remove instant once we add time mocking
+    pub fn record(&mut self, bytes: u64, instant: Instant) {
+        self.bytes_per_min += bytes;
+        self.events.push_back(Event { instant, bytes })
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs:31]: Good idea.!

File: chain/network/src/peer/peer_actor.rs [hunk 6]
```diff
@@ -1010,15 +1013,18 @@ impl Handler<QueryPeerStats> for PeerActor {
...
-            received_bytes_per_sec: self.tracker.received_bytes.bytes_per_min() / 60,
-            sent_bytes_per_sec: self.tracker.sent_bytes.bytes_per_min() / 60,
+            received_bytes_per_sec: received.bytes_per_min / 60,
+            sent_bytes_per_sec: sent.bytes_per_min / 60,
             is_abusive: self.is_abusive(),
-            message_counts: (
-                self.tracker.sent_bytes.count_per_min(),
-                self.tracker.received_bytes.count_per_min(),
-            ),
+            message_counts: (sent.count_per_min, received.count_per_min),
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs:1027]: That's a good point. I was thinking about doing it in another PR. Just excluded it from here  for sake of simplicity.

File: chain/network/src/peer/peer_actor.rs [hunk 7]
```diff
@@ -1010,15 +1013,18 @@ impl Handler<QueryPeerStats> for PeerActor {
...
+        let sent = self.tracker.sent_bytes.minute_stats(now);
+        let received = self.tracker.received_bytes.minute_stats(now);
+
         PeerStatsResult {
             chain_info: self.chain_info.clone(),
-            received_bytes_per_sec: self.tracker.received_bytes.bytes_per_min() / 60,
-            sent_bytes_per_sec: self.tracker.sent_bytes.bytes_per_min() / 60,
+            received_bytes_per_sec: received.bytes_per_min / 60,
+            sent_bytes_per_sec: sent.bytes_per_min / 60,
             is_abusive: self.is_abusive(),
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs:1026]: Fixed

File: chain/network/src/peer/peer_actor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs]: I added it to list of things to do https://github.com/near/nearcore/issues/5423

File: chain/network/src/peer/rate_counter.rs [hunk 8]
```diff
@@ -1,86 +1,89 @@
...
-    /// Time we created the entry.
-    recorded: SystemTime,
 }
 
-/// A rate counter tracks number of transfers, the amount of data exchanged and the rate of transfer
-/// over the last minute.
-pub struct RateCounter {
-    entries: VecDeque<Entry>,
-    bytes_sum: u64,
+// Measure per minute transfer stats
```
Comment (matklad) [inline chain/network/src/peer/rate_counter.rs:9]: Three `///` please :-)

File: chain/network/src/peer/peer_actor.rs [hunk 9]
```diff
@@ -1009,15 +1011,18 @@ impl Handler<QueryPeerStats> for PeerActor {
...
-            sent_bytes_per_sec: self.tracker.sent_bytes.bytes_per_min() / 60,
-            is_abusive: self.is_abusive(),
-            message_counts: (
-                self.tracker.sent_bytes.count_per_min(),
-                self.tracker.received_bytes.count_per_min(),
-            ),
+            received_bytes_per_sec: received.bytes_per_min / 60,
+            sent_bytes_per_sec: sent.bytes_per_min / 60,
+            is_abusive: self.is_abusive(now),
+            message_counts: (sent.count_per_min, received.count_per_min),
```
Comment (matklad) [inline chain/network/src/peer/peer_actor.rs:1025]: Ultraminor nit, but I'd 

```
        let sent = self.tracker.sent_bytes.minute_stats(now);
        let received = self.tracker.received_bytes.minute_stats(now);
        let is_abusive = self.is_abusive(now);
```

to keep all statements where we look at time bunched up.

Comment (matklad) [reply chain/network/src/peer/peer_actor.rs:1025]: Actually, now that I wrote it that way, I wonder if `is_abusive` needs to be a method on `self`? Like, the first thing it does is re-doing the computation we did on the previous line. That's the case where

```rust
let is_abusive = {
   // some comment for the block
   computation
};
```

pattern might we a win over an extracted function.

Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs:1025]: Yes, inlining is a good idea. I updated the PR.

File: chain/network/src/peer/rate_counter.rs [hunk 10]
```diff
@@ -1,86 +1,93 @@
...
-pub struct RateCounter {
-    entries: VecDeque<Entry>,
-    bytes_sum: u64,
+/// Represents all events which happened in last minute.
+#[derive(Default)]
+pub struct TransferStats {
+    /// We keep list of entries not older than 1m.
+    events: VecDeque<Event>,
+    /// Sum of bytes for all entries.
+    bytes_per_min: u64,
```
Comment (mm-near) [inline chain/network/src/peer/rate_counter.rs:18]: nit; I'd not call it 'per_min' - maybe bytes_in_events? or something similar?

File: chain/network/src/peer/rate_counter.rs [hunk 11]
```diff
@@ -1,86 +1,93 @@
...
-    pub fn increment(&mut self, bytes: u64) {
-        let now = SystemTime::now();
-        self.entries.push_back(Entry { bytes, recorded: now });
-        self.bytes_sum += bytes;
-        self.truncate(now);
-    }
+/// Represents cumulative stats per minute.
+#[derive(Eq, PartialEq, Debug)]
+pub struct MinuteStats {
+    /// Byres per minute.
```
Comment (mm-near) [inline chain/network/src/peer/rate_counter.rs:24]: s/Byres/Bytes (same 2 lines below)

File: chain/network/src/peer/rate_counter.rs [hunk 12]
```diff
@@ -1,86 +1,93 @@
...
+    pub bytes_per_min: u64,
+    /// Byres per minute.
+    pub count_per_min: usize,
+}
 
-    pub fn bytes_per_min(&self) -> u64 {
-        self.bytes_sum
+impl TransferStats {
```
Comment (mm-near) [inline chain/network/src/peer/rate_counter.rs:30]: nit: I'd either keep the name 'rateCounter' (or consider renaming the file)

File: chain/network/src/peer/rate_counter.rs [hunk 13]
```diff
@@ -1,86 +1,93 @@
...
+    /// Record event at given `instant` with `bytes` bytes.
+    pub fn record(&mut self, bytes: u64, instant: Instant) {
+        self.bytes_per_min += bytes;
```
Comment (mm-near) [inline chain/network/src/peer/rate_counter.rs:33]: what if instant is far in the future?

File: chain/network/src/peer/rate_counter.rs [hunk 14]
```diff
@@ -1,86 +1,93 @@
...
+        MinuteStats { bytes_per_min: self.bytes_per_min, count_per_min: self.events.len() }
     }
 
-    fn truncate(&mut self, now: SystemTime) {
-        // Remove entries older than 1m.
-        while !self.entries.is_empty() && self.entries.front().unwrap().recorded < now - MINUTE {
-            self.bytes_sum -= self.entries.pop_front().unwrap().bytes;
+    /// Remove entries older than 1m.
+    fn remove_old_entries(&mut self, instant: Instant) {
+        while let Some(event) = self.events.pop_front() {
```
Comment (mm-near) [inline chain/network/src/peer/rate_counter.rs:46]: this assumes that objects in the events queue are in order, right?

**Review by mm-near — CHANGES_REQUESTED** [review]

File: chain/network/src/peer/rate_counter.rs (hunk 10 above)
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs:18]: I renamed to `total_bytes_in_events`.

File: chain/network/src/peer/rate_counter.rs (hunk 12 above)
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs:30]: @mm-near I renamed the file to `transfer-stats`. I rewrote the whole file, it's better not to keep using the name due to license.

File: chain/network/src/peer/rate_counter.rs (hunk 13 above)
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs:33]: @mm-near I renamed it to `now`, and updated description.

File: chain/network/src/peer/rate_counter.rs (hunk 14 above)
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs]: @mm-near Yes, that's how it works. I updated the documentation to make code easier to read.

File: chain/network/src/peer/peer_actor.rs [hunk 15]
```diff
@@ -52,7 +52,7 @@ type WriteHalf = tokio::io::WriteHalf<tokio::net::TcpStream>;
 
 /// Maximum number of messages per minute from single peer.
 // TODO: current limit is way to high due to us sending lots of messages during sync.
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs]: Ok, Added.

File: chain/network/src/peer/rate_counter.rs (hunk 14 above)
Comment (pmnoxx) [reply chain/network/src/peer/rate_counter.rs]: Yes, I updated docs.

**Comment by pmnoxx** [comment]
```
@mm-near: chain/network/ chain/network-primitives/
```
@mm-near Can you take a look?

**Comment by pmnoxx** [comment]
Closing this PR in favor of https://github.com/near/nearcore/pull/6020

**Pull request closed without merging**